From 4914063dffb07fc2cdd393e2242dd521e2bd8026 Mon Sep 17 00:00:00 2001
From: Rui Loura <rui@joyent.com>
Date: Mon, 9 Jul 2018 20:39:41 +0000
Subject: [PATCH] TRITON-574 sdcadm platform assign is broken with rack aware
 networking Reviewed by: Cody Peter Mello <cody.mello@joyent.com>

---
 CHANGES.md      |  4 ++++
 lib/platform.js | 47 ++++++++++++++++++++++++++++++++++++++++++++++-
 package.json    |  2 +-
 3 files changed, 51 insertions(+), 2 deletions(-)

diff --git a/CHANGES.md b/CHANGES.md
index e1de9c7..62d1bd1 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -10,6 +10,10 @@
 
 # sdcadm Changelog
 
+## 1.20.0
+
+- TRITON-574 sdcadm platform assign is broken with rack aware networking
+
 ## 1.19.1
 
 - TRITON-572 sdcadm post-setup cloudapi should let sapi determine instance networks
diff --git a/lib/platform.js b/lib/platform.js
index 548b5a4..28e3420 100644
--- a/lib/platform.js
+++ b/lib/platform.js
@@ -1066,6 +1066,7 @@ Platform.prototype.assign = function assign(opts, callback) {
     var serverRecs = [];
     var assignServers = [];
     var uuids = [];
+    var adminTags = [];
     var headnode;
     var progress = self.progress;
 
@@ -1074,6 +1075,19 @@ Platform.prototype.assign = function assign(opts, callback) {
     var errs = [];
 
 
+    /*
+     * The booter(dhcpd) zone maintains a set of bootparams for each admin mac
+     * address of each server(CN).  The bootparams specify which platform image
+     * is assigned to a given CN.
+     *
+     * This step updates the bootparams on the dhcpd(booter) zone.  First it
+     * gathers all of the admin NICs from both the 'admin' network, and the
+     * 'admin' network pool that belong to servers(CNs).  Then it retrieves
+     * the associated mac addresses which are then passed as arguments to a
+     * command line tool("booter bootparams") on the booter zone.  This tool
+     * will update each server's boot params (as stored in the booter zone) to
+     * specify the new platform image for each given admin mac address.
+     */
     function updateBooterCache(servers, cb) {
         var macs;
         var serveruuids = servers.map(function (server) {
@@ -1083,10 +1097,41 @@ Platform.prototype.assign = function assign(opts, callback) {
         progress('Updating booter cache for servers');
 
         vasync.pipeline({funcs: [
+            function getAdminNicTags(_, next) {
+                self.sdcadm.napi.listNetworkPools({name: 'admin'},
+                    function (err, pools) {
+
+                    if (err) {
+                        next(new errors.SDCClientError(err, 'napi'));
+                        return;
+                    }
+
+                    if (pools.length > 1) {
+                        progress('Multiple network pools named "admin", '
+                            + 'using %s', pools[0].uuid);
+                    }
+
+                    /*
+                     * The presence of an admin network pool is optional.
+                     */
+                    if (pools.length === 0) {
+                        next();
+                        return;
+                    }
+
+                    adminTags = pools[0].nic_tags_present;
+                    next();
+                });
+
+            },
             function (_, next) {
+                if (adminTags.indexOf('admin') === -1) {
+                    adminTags.push('admin');
+                }
+
                 var listOpts = {
                     belongs_to_type: 'server',
-                    nic_tags_provided: 'admin'
+                    nic_tags_provided: adminTags
                 };
                 if (!opts.all) {
                     listOpts.belongs_to_uuid = serveruuids;
diff --git a/package.json b/package.json
index f25ff4b..32b5229 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "sdcadm",
   "description": "Administer a SmartDataCenter (SDC) standup",
-  "version": "1.19.1",
+  "version": "1.20.0",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
-- 
2.21.0

